contracts/Settlement/SettlementLogic.sol

```diff
@@ -446,52 +528,148 @@ contract SettlementLogic is ISettlement, SettlementStorage {
     }
 
-    function _checkRelayerFee(uint256 fee, address fromToken) internal view {
-        uint256 feeInRbtc = fee;
+    function _checkRelayerFee(
+        address fromToken,
+        uint256 orderSize,
+        uint256 amountToFill,
+        bool isSpot
+    ) internal view returns (uint256 relayerFee) {
+        uint256 estOrderFee = amountToFill.mul(relayerFeePercent).div(10**20);
+        uint256 minFeeAmount = isSpot ? minSwapOrderTxFee : minMarginOrderTxFee;
+        uint256 minFeeAmountInToken = minFeeAmount;
+
         if (fromToken != WRBTC_ADDRESS) {
             address[] memory path = sovrynSwapNetwork.conversionPath(
-                fromToken,
-                WRBTC_ADDRESS
+                WRBTC_ADDRESS,
+                fromToken
+            );
+            minFeeAmountInToken = sovrynSwapNetwork.rateByPath(
+                path,
+                minFeeAmount
             );
-            feeInRbtc = sovrynSwapNetwork.rateByPath(path, fee);
         }
-        require(
-            feeInRbtc > minFee,
-            "Order amount is too low to pay the relayer fee"
+
+        if (estOrderFee < minFeeAmountInToken) {
+            require(amountToFill > minFeeAmountInToken, "Order amount is too low to pay the relayer fee");
+            require(
+                orderSize > minFeeAmountInToken,
+                "Order amount is too low to pay the relayer fee"
+            );
+
+            relayerFee = minFeeAmountInToken;
+        } else {
+            uint256 fillAmountInRBtc = amountToFill;
+            uint256 minFillingAmount = isSpot
+                ? minSwapOrderSize
+                : minMarginOrderSize;
+            if (fromToken != WRBTC_ADDRESS) {
+                address[] memory pathToRbtc = sovrynSwapNetwork.conversionPath(
+                    fromToken,
+                    WRBTC_ADDRESS
+                );
+                fillAmountInRBtc = sovrynSwapNetwork.rateByPath(
+                    pathToRbtc,
+                    amountToFill
+                );
+            }
+
+            require(
+                fillAmountInRBtc >= minFillingAmount,
+                "Filling amount is too low"
+            );
+            relayerFee = estOrderFee;
+        }
+    }
+
+    function _depositOrderAsset(
+        address owner,
+        address assetAddress,
+        uint256 amount
+    ) internal {
+        if (assetAddress == WRBTC_ADDRESS) {
+            IWrbtcERC20 wrbtc = IWrbtcERC20(WRBTC_ADDRESS);
+            require(balanceOf[owner] >= amount, "insufficient-balance");
+            balanceOf[owner] -= amount;
+            wrbtc.deposit{value: amount}();
+        } else {
+            IERC20(assetAddress).safeTransferFrom(owner, address(this), amount);
+        }
+    }
+
+    function _checkWithdrawalOnCancel(
+        address owner,
+        address token,
+        uint256 amount
+    ) internal {
+        if (token == WRBTC_ADDRESS && balanceOf[owner] >= amount) {
+            withdraw(amount);
```

Comment:
`_checkWithdrawalOnCancel` is called from the functions with meta-transaction elements, and `withdraw` relies on `msg.sender` internally. In this particular case it works fine, but it is one step from critical issue.

Reply:
why are you passing the `owner` here at all if in withdraw you're using msg.sender?

Reply:
Removed owner: https://github.com/DistributedCollective/LimitOrder/pull/7/commits/1a4c7951182a83f7514aa16f6fdc2f9dd683582a